### PR TITLE
Integration Test Configuration

### PR DESCRIPTION
**Pull request by nholland94**
The PR includes a new test runner which is intended to be used by CI and devs alike for dispatching integration tests. The configuration provided in the script is used to render the CI configuration. The script supports selecting or blacklisting test configurations using a simple syntax (`<profile>:<test>` where `*` is a wildcard).

### CODE REVIEW

File: scripts/test.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    def run_cmd(cmd, on_failure):
+        def do():
+            if subprocess.call(['bash', '-c', cmd], stdout=sys.stdout, stderr=sys.stderr) != 0:
+                on_failure()
+        wet('$ ' + cmd, do)
+
+    coda_exe_path = 'src/app/cli/src/coda.exe'
+    coda_build_path = './_build/default'
+    coda_exe = os.path.join(coda_build_path, coda_exe_path)
```
Comment (cmr) [inline scripts/test.py]: this script can't be run from inside `src`. I don't know how much of that people do, but it's the default configuration of `make build`. It'd be nice if this script were robust to either.

File: scripts/test.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    output_file.write(rendered)
+    output_file.close()
+
+def main():
+    actions = {
+        'run': run,
+        'render': render
+    }
+
+    root_parser = argparse.ArgumentParser(description='Coda integration test runner/configurator')
```
Comment (cmr) [inline scripts/test.py]: The help strings aren't very useful. A brief note about the filter language would be useful. Additionally, a command listing all the possibilities for both halves of the pattern would be handy - so you don't have to read the source to figure out the ways this program can be used.

File: scripts/test.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            for (profile_pat, test_pat) in whitelist_patterns
+        )
+        blacklisted = any(
+            test_pattern(profile_pat, profile) and test_pattern(test_pat, test)
+            for (profile_pat, test_pat) in blacklist_patterns
+        )
+        return whitelisted and not(blacklisted)
+
+    # too complex to represent as a list expression in python
+    result = {}
```
Comment (cmr) [inline scripts/test.py]: use `result = collections.defaultdict(list)` instead

File: scripts/test.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    for (profile,tests) in test_permutations.items():
+        for test in tests:
+            if keep(profile, test):
+                result[profile] = result[profile] if profile in result else []
```
Comment (cmr) [inline scripts/test.py]: (then this line will be unnecessary)

File: scripts/test.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    render_parser.add_argument('jinja_file')
+
+    args = root_parser.parse_args()
+    if not(hasattr(args, 'action')):
+        print('Unrecognized action')
+        root_parser.print_help()
+        exit(1)
+    actions[args.action](args)
+
+main()
```
Comment (cmr) [inline scripts/test.py]: `if __name__ == "__main__":` guard this, please.

File: Makefile [hunk 6]
```diff
@@ -214,33 +214,7 @@ codaslim:
 ## Tests
 
 render-circleci:
-	cd .circleci; python2 render.py > config.yml
-
-check-render-circleci:
```
Comment (cmr) [inline Makefile]: can you add this check back?

File: scripts/test.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    print('Preparing to run the following tests:')
+    for (profile, tests) in test_permutations.items():
+        print('- %s:' % profile)
+        for test in tests:
+            print('  - %s' % test)
+
+    timestamp = int(time.time())
+    print('======================================')
+    print('=============', timestamp, '=============')
```
Comment (cmr) [inline scripts/test.py]: this is printing as a tuple.

Comment (nholland94) [reply scripts/test.py]:
```
======================================
============= 1552440690 =============
======================================
```
looks like that for me

Comment (cmr) [reply scripts/test.py]: oh, it's py2 vs py3.

Comment (nholland94) [reply scripts/test.py]: I'll change it to do string concatenation since apparently that's different across versions.

File: scripts/test.py [hunk 8]
```diff
@@ -108,13 +107,16 @@ def do():
...
+    coda_exe_path = 'app/cli/src/coda.exe'
+    if not(os.path.exists('dune-project')):
+        coda_exe_path = os.path.join('src', coda_exe_path)
     coda_build_path = './_build/default'
     coda_exe = os.path.join(coda_build_path, coda_exe_path)
 
     jq_filter = '.level=="Warn" or .level=="Error" or .level=="Faulty_peer" or .level=="Fatal"'
 
-    test_permutations = filter_test_permutations([args.test_pattern], [args.blacklist_pattern])
+    print(args)
```
Comment (cmr) [inline scripts/test.py]: leftover?

File: scripts/test.py [hunk 9]
```diff
@@ -169,6 +171,7 @@ def render(args):
     (output_file, ext) = os.path.splitext(args.jinja_file)
     assert ext == '.jinja'
 
+    print(ci_blacklist)
```
Comment (cmr) [inline scripts/test.py]: leftover?

File: scripts/test.py [hunk 10]
```diff
@@ -183,30 +186,61 @@ def render(args):
...
-    subparsers = root_parser.add_subparsers(help='command help')
-
-    run_parser = subparsers.add_parser('run', help='run help')
+    root_parser = argparse.ArgumentParser(description='Coda integration test runner/configurator.')
+    subparsers = root_parser.add_subparsers(help='subcommands')
+
+    run_parser = subparsers.add_parser(
+        'run',
+        description='''
+            Build and run integration tests. Filters can be provided for
```
Comment (cmr) [inline scripts/test.py]: <3

File: scripts/test.py [hunk 11]
```diff
@@ -182,9 +180,13 @@ def render(args):
         test_permutations=test_permutations
     )
 
-    output_file = open(output_file, 'w')
-    output_file.write(rendered)
-    output_file.close()
+    if args.check:
+        with open(output_file, 'r') as file:
+            if file.read() != rendered:
+                fail('circle CI configuration was not rendered correctly')
```
Comment (cmr) [inline scripts/test.py]: I'd prefer if this string said `circle CI configuration is out of date, re-render it`

**Pull request merged**